atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/CheckAndSetCompatibility.java

```diff
@@ -16,64 +16,95 @@
+     * If it is not true, behaviour is undefined.
      */
-    public boolean supportsDetailOnFailure() {
-        return supportsDetailOnFailure;
-    }
+    boolean supportsDetailOnFailure();
 
     /**
      *  If true, on CAS or PUE failure other than a {@link CheckAndSetException} or a
      *  {@link KeyAlreadyExistsException}, the values may or may not have been persisted but the state is guaranteed to
      *  be consistent: any subsequent reads will be repeatable. If false, the value may have been persisted
      *  in a way that subsequent reads are not repeatable: if a read before the operation could have returned values
      *  from a set S, it can non-deterministically return a value from S U {newValue} afterwards.
+     *
+     * This method should only be called if {@link CheckAndSetCompatibility#supportsCheckAndSetOperations()} is true.
+     * If it is not true, behaviour is undefined.
      */
-    public boolean consistentOnFailure() {
-        return consistentOnFailure;
-    }
+    boolean consistentOnFailure();
 
-    public static CheckAndSetCompatibility intersect(Stream<CheckAndSetCompatibility> compatibilities) {
+    static CheckAndSetCompatibility intersect(Stream<CheckAndSetCompatibility> compatibilities) {
         Set<CheckAndSetCompatibility> presentCompatibilities = compatibilities.collect(Collectors.toSet());
+
+        boolean supported =
+                presentCompatibilities.stream().allMatch(CheckAndSetCompatibility::supportsCheckAndSetOperations);
+        if (!supported) {
+            return Unsupported.INSTANCE;
+        }
         boolean detail = presentCompatibilities.stream().allMatch(CheckAndSetCompatibility::supportsDetailOnFailure);
         boolean consistency = presentCompatibilities.stream().allMatch(CheckAndSetCompatibility::consistentOnFailure);
 
-        return Stream.of(CheckAndSetCompatibility.values())
-                .filter(compatibility -> compatibility.supportsDetailOnFailure() == detail)
-                .filter(compatibility -> compatibility.consistentOnFailure() == consistency)
-                .findFirst()
-                .orElseThrow(() -> new SafeIllegalArgumentException("min requires at least 1 element, but 0 provided"));
+        return supportedBuilder()
+                .supportsDetailOnFailure(detail)
+                .consistentOnFailure(consistency)
+                .build();
+    }
+
+    static CheckAndSetCompatibility unsupported() {
+        return Unsupported.INSTANCE;
+    }
+
+    static ImmutableSupported.Builder supportedBuilder() {
+        return ImmutableSupported.builder();
+    }
+
+    enum Unsupported implements CheckAndSetCompatibility {
+        INSTANCE;
+
+        @Override
+        public boolean supportsCheckAndSetOperations() {
+            return false;
+        }
+
+        @Override
+        public boolean supportsDetailOnFailure() {
+            throw new SafeIllegalStateException(
```

Comment:
Is this illegal state or illegal argument? Maybe even UnsupportedOperation tbh?

Reply:
Let's use UnsupportedOperation, yeah.